pkcs11/work.go

```diff
@@ -26,68 +35,193 @@ type Work struct {
 		}
-	}(workerCtx)
+	}(reqCtx)
 
 	for {
 		select {
 		case <-workerCtx.Done():
-			// Case 1: Worker stopped either due to context cancelled or worker timed out.
-			// This case is to avoid worker being stuck in a blocking call or a deadlock scenario.
-			log.Printf("%s: worker stopped", worker.String())
+			// Case 1: Worker stopped or cancelled request.
+			// The client is still waiting for a response, so return on error channel.
+			worker.TotalTimeout.Inc()
 			cancel()
-			w.sendResponse(nil)
+			w.work.errChan <- errors.New("worker cancelled request")
 			return
-		case resp := <-signerRespCh:
-			// Case 2: Received response. It could either be a pkcs11 timeout or thr worker was able to get a signer
-			// from the signer pool.
-			if resp.signer == nil || resp.err != nil {
-				worker.TotalTimeout.Inc()
-				log.Printf("%s: error fetching signer %v", worker.String(), resp.err)
-				w.work.respChan <- nil
-				cancel()
-				return
-			}
-			worker.TotalProcessed.Inc()
-			w.sendResponse(resp.signer)
+		case <-reqCtx.Done():
+			// Case 2: HSM/PKCS11 request timed out.
+			// The client is still waiting for a response in this case, so return on error channel.
+			worker.TotalTimeout.Inc()
+			w.work.errChan <- errors.New("hsm request timed out")
+			return
+		case <-w.work.stop:
+			// Case 3: Client cancelled the request.
+			// In this case we no longer need to process the signing request & we should clean up signer if assigned & return.
+			worker.TotalTimeout.Inc()
 			cancel()
 			return
-		case _, ok := <-w.work.respChan:
-			// Case 3: Check for current state of respChan. If the client request is cancelled, the client
-			// will close the respChan. In that case, we should cancel reqCtx & return to avoid extra processing.
-			if !ok {
-				log.Printf("%s: worker request timed out, client cancelled request", worker.String())
-				cancel()
+		case sResp := <-signerResp:
+			pt = time.Since(pStart).Nanoseconds() / time.Microsecond.Nanoseconds()
+			// Case 4: Received signer from signer pool. We need to sign the request & send the response. Before we send the
+			// response, we should ensure client is still waiting for the response.
+			if sResp.err != nil {
 				worker.TotalTimeout.Inc()
+				w.work.errChan <- errors.New("client request timed out, skip signing request")
 				return
 			}
+			worker.TotalProcessed.Inc()
+
+			hStart = time.Now()
+			switch w.work.method {
+			case "GetSSHCertSigningKey", "GetX509CACert", "GetBlobSigningPublicKey":
+				go w.work.signerData.getData(reqCtx, sResp.signer, w.work.pool, dataCh, errCh)
+			case "SignSSHCert", "SignX509Cert", "SignBlob":
+				go w.work.signerData.signData(reqCtx, sResp.signer, w.work.pool, dataCh, errCh)
+			}
+		case resp := <-dataCh:
+			ht = time.Since(hStart).Nanoseconds() / time.Microsecond.Nanoseconds()
+			// Case 5: HSM has completed the signing operation & we need to send response back to client.
+			w.work.respChan <- resp
```

Comment:
The worker (routine) will be blocked at `respChan` or `errChan` if the `ctx` at signer.go L116 gets timeout earlier.

Ref: https://github.com/theparanoids/crypki/pull/168/files#diff-1c28a02b075d9ba8dd534726c1bf7c8f4af1bcfb853ccab47cabcce948cac257R116

Reply:
Approach to fix: https://github.com/theparanoids/crypki/pull/173